pkg/controller/common/rollout/workloads/cloneset_controller.go

```diff
@@ -1,32 +1,199 @@
+		for i, r := range c.rolloutSpec.RolloutBatches {
+			batchSize, _ := intstr.GetValueFromIntOrPercent(&r.Replicas, int(*cloneSetSize), false)
+			upgradedReplicasStatus -= batchSize
+			if upgradedReplicasStatus <= 0 {
+				currentBatch = i
+				break
+			}
+		}
+		klog.InfoS("locate the batch to upgrade with automatic rollout", "current batch", currentBatch)
+	} else {
+		// the rollout is manual, we verify that the UpgradedReplicas lands in the batch we should upgrade
+		activePartition := int(*c.rolloutSpec.BatchPartition)
+		// calculate the lower bound of the possible upgradeReplicas just before current batch
+		for i, r := range c.rolloutSpec.RolloutBatches {
+			if i < activePartition {
+				batchSize, _ := intstr.GetValueFromIntOrPercent(&r.Replicas, int(*cloneSetSize), false)
+				upgradeReplicas += batchSize
+			}
+		}
+		// the recorded number should be at least as much as the all the pods before the active batch
+		if upgradeReplicas > upgradedReplicasStatus {
+			klog.ErrorS(fmt.Errorf("the upgraded replica in the status is too small"), "upgraded replicas status",
+				upgradedReplicasStatus, "replicas in all the previous batches", upgradeReplicas)
+		}
+		// calculate the upper bound with the current batch
+		batchSize, _ := intstr.GetValueFromIntOrPercent(&c.rolloutSpec.RolloutBatches[activePartition].Replicas,
+			int(*cloneSetSize), false)
+		upgradeReplicas += batchSize
+		// the recorded number should be not as much as the all the pods including the active batch
+		if upgradeReplicas < upgradedReplicasStatus {
+			klog.ErrorS(fmt.Errorf("the upgraded replica in the status is too large"), "upgraded replicas status",
+				upgradedReplicasStatus, "replicas in the batches including the current batch", upgradeReplicas)
+		}
+	}
+	err := c.rolloutStatus.StateTransition(v1alpha1.BatchRolloutApprovedEvent)
+	if err != nil {
+		// panic for now
+		panic(err.Error())
+	}
+	return *c.rolloutStatus
+
 }
 
 // Finalize makes sure the Cloneset is all upgraded and
-func (c *CloneSetController) Finalize() error {
+func (c *CloneSetController) Finalize(ctx context.Context) v1alpha1.RolloutStatus {
+	if c.fetchCloneSet(ctx) != nil {
+		return *c.rolloutStatus
+	}
+	// mark the rollout finalized
+	c.recorder.Event(c.parentController, event.Normal("Finalized", "Rollout resource are finalized"))
+	err := c.rolloutStatus.StateTransition(v1alpha1.RollingFinalizedEvent)
+	if err != nil {
+		// panic for now
+		panic(err.Error())
+	}
+	return *c.rolloutStatus
+}
+
+// The rest are helper functions
+func (c *CloneSetController) fetchCloneSet(ctx context.Context) error {
+	// get the cloneSet
+	workload := kruise.CloneSet{}
+	err := c.client.Get(ctx, c.workloadNamespacedName, &workload)
+	if err != nil {
+		klog.CalculateMaxSize()
+		if !apierrors.IsNotFound(err) {
+			c.rolloutStatus.RolloutRetry(err.Error())
+			c.recorder.Event(c.parentController, event.Warning("Failed to get the Cloneset", err))
+		}
+		c.rolloutStatus.RolloutFailed(err.Error())
```

Comment:
if it's not found, why don't we create it? There can be a case that we use rollout to create or scale up.

1. create: there will no source workload, and the target workload is also not exist.
2. scale up: target workload will equal with source worklaod and only replicas diff

Reply:
The contract between the rollout controller and the outside (either application/appDepolyment/rolloutTrait) is that the outside controller creates the resources. 

I think we can retry on the NotFound case though.